### PR TITLE
rgw: work around curl_multi_wait bug with non-blocking reads

### PR DESCRIPTION
**Pull request by cbodley**
Fixes: http://tracker.ceph.com/issues/15915
Fixes: http://tracker.ceph.com/issues/16695

Signed-off-by: Casey Bodley <cbodley@redhat.com>

### CODE REVIEW

**Comment by cbodley** [comment]
this made it through teuthology with no new failures: http://pulpito.ceph.com/cbodley-2016-09-02_15:45:38-rgw-wip-16695---basic-mira

**Comment by oritwas** [comment]
this means we do a read every time do_curl_wait is called, correct?
can we detect what curl version is running and only for older versions call read without checking wait_fd.revents?

**Comment by cbodley** [comment]
@oritwas yeah, that was my first inclination. i wasn't going to require this of the original contributor in #10467, but i agree that it's worth doing

**Comment by cbodley** [comment]
tested and verified using unittest_http_manager and an old version of libcurl 7.29 located at /home/cbodley/local-curl/lib/libcurl.so

when run against master, the bug leads to deadlock:
```
$ LD_PRELOAD=/home/cbodley/local-curl/lib/libcurl.so ./unittest_http_manager
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from HTTPManager
[ RUN      ] HTTPManager.SignalThread
^C
```
when run against this branch, the bug is detected and the workaround avoids deadlock:
```
$ LD_PRELOAD=/home/cbodley/local-curl/lib/libcurl.so ./unittest_http_manager
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from HTTPManager
[ RUN      ] HTTPManager.SignalThread
2016-09-09 16:55:25.457454 7eff9c96e900  0 WARNING: detected a version of libcurl which contains a bug in curl_multi_wait(). enabling a workaround that may degrade performance slightly. please upgrade to a more recent version of libcurl.
[       OK ] HTTPManager.SignalThread (71 ms)
[----------] 1 test from HTTPManager (71 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (71 ms total)
[  PASSED  ] 1 test.
```
and with my system's libcurl 7.43, the workaround is not applied and the test passes normally:
```
$ ./unittest_http_manager
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from HTTPManager
[ RUN      ] HTTPManager.SignalThread
[       OK ] HTTPManager.SignalThread (74 ms)
[----------] 1 test from HTTPManager (74 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (74 ms total)
[  PASSED  ] 1 test.
```

File: src/rgw/rgw_http_client.cc [hunk 1]
```diff
@@ -418,6 +418,16 @@ int RGWHTTPTransceiver::send_data(void* ptr, size_t len)
 }
 
 
+static int clear_signal(int fd)
+{
+  uint32_t buf;
+  int ret = ::read(fd, &buf, sizeof(buf));
```
Comment (yehudasa) [inline src/rgw/rgw_http_client.cc]: @cbodley need to cast to void * like we had originally? might be that needed for clearing a warning

File: src/rgw/rgw_http_client.cc [hunk 2]
```diff
@@ -420,8 +420,10 @@ int RGWHTTPTransceiver::send_data(void* ptr, size_t len)
 
 static int clear_signal(int fd)
 {
-  uint32_t buf;
-  int ret = ::read(fd, &buf, sizeof(buf));
+  // since we're in non-blocking mode, we can try to read a lot more than
+  // one signal from signal_thread() to avoid later wakeups
+  std::array<char, 256> buf;
+  int ret = ::read(fd, buf.data(), buf.size());
   if (ret < 0) {
```
Comment (yehudasa) [inline src/rgw/rgw_http_client.cc]: @cbodley should we handle EAGAIN?

Comment (yehudasa) [inline src/rgw/rgw_http_client.cc]: e.g., clear it?

**Comment by yehudasa** [comment]
@cbodley see my comments, otherwise lgtm

**Comment by cbodley** [comment]
@yehudasa thanks for the feedback, updated `clear_signal()` to preserve the (void *) cast and return 0 on EAGAIN

**Pull request merged**